File: src/main/year2015/day01/Day01.java
package year2015.day01;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.List;
import java.util.regex.Pattern;

public class Day01 {
    
    private final List<String> input;

    public Day01(String text) throws IOException {
        String day = "01";
        String year = "2015";
        String filePath = "src/resources/" + year + "/" + day + "/" + text;
        input = Files.readAllLines(Paths.get(filePath));
    }
    public int part1() {
        return (int) input.stream()
                .filter(c -> !Pattern.compile("ab|cd|pq|xy").matcher(c).find())
                .filter(c -> Pattern.compile("(.*[aeiou]){3}").matcher(c).find())
                .filter(c -> Pattern.compile("(.)\\1").matcher(c).find())
                .count();
    }

    public int part2() {
        return (int) input.stream()
                .filter(c -> Pattern.compile("(..).*\\1").matcher(c).find())
                .filter(c -> Pattern.compile("(.).\\1").matcher(c).find())
                .count();
    }
}

File: src/main/year2019/day01/Day01.kt
package main.year2019.day01

import java.io.File

class Day01(text : String = "input.txt") {

    private val day = "01"
    private val year = "2019"

    private val input = File("src/resources/$year/$day/$text").readLines().toList().map { it.toInt() }.toIntArray()

    fun part1(): Int = input.sumOf { it / 3 - 2 }
    fun part2(): Int = input.sumOf { fuel(it) }

    //recursive to get the amount of fuel needed for the fuel
    private fun fuel(it: Int) : Int = if (it / 3 - 2 > 0) it / 3 - 2 + fuel(it / 3 - 2) else 0

}

fun main() {
    val day = Day01()
    println(day.part1())
    println(day.part2())
}

